### PR TITLE
CNFCERT-239:Expand IsCertified test to include Helm Chart certificati…

### PR DESCRIPTION
**Pull request by aabughosh**
…on catalog API check

### CODE REVIEW

File: pkg/config/configsections/common.go [hunk 1]
```diff
@@ -99,6 +99,7 @@ type TestTarget struct {
 	ExcludeContainersFromMultusConnectivityTests []ContainerIdentifier `yaml:"excludeContainersFromMultusConnectivityTests" json:"excludeContainersFromMultusConnectivityTests"`
 	// Operator is the list of operator objects that needs to be tested.
 	Operators []Operator `yaml:"operators,omitempty"  json:"operators,omitempty"`
+	Helm      []Helm     `yaml:"helm" json:"helm"`
```
Comment (edcdavid) [inline pkg/config/configsections/common.go:102]: Helm      []Helm -> HelmCharts      []HelmChart

File: pkg/config/configsections/helm.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+//
+// You should have received a copy of the GNU General Public License along
+// with this program; if not, write to the Free Software Foundation, Inc.,
+// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+package configsections
+
+type Helm struct {
+	Version string
+	Chart   string
```
Comment (edcdavid) [inline pkg/config/configsections/helm.go:21]: type HelmChart struct {
	Version string
	name   string
}

File: pkg/config/config.go [hunk 3]
```diff
@@ -237,6 +238,7 @@ func (env *TestEnvironment) doAutodiscover() {
 	env.DeploymentsUnderTest = env.Config.DeploymentsUnderTest
 	env.StateFulSetUnderTest = env.Config.StateFulSetUnderTest
 	env.OperatorsUnderTest = env.Config.Operators
+	env.Helmcharts = env.Config.Helm
```
Comment (edcdavid) [inline pkg/config/config.go:241]: Helmcharts -> HelmchartsUnderTest

File: test-network-function/identifiers/identifiers.go [hunk 4]
```diff
@@ -431,6 +436,15 @@ with no resourceNames under its rules.`),
 		BestPracticeReference: bestPracticeDocV1dot2URL + " Section 6.2.12 and Section 6.3.3",
 	},
 
+	TestHelmIsCertifiedIdentifier: {
+		Identifier:  TestHelmIsCertifiedIdentifier,
+		Type:        normativeResult,
+		Remediation: `Ensure that your helms list has passed Red Hat's helm Certification Program (OCP).`,
```
Comment (edcdavid) [inline test-network-function/identifiers/identifiers.go:442]: Ensure that helm charts passed the Red Hat's helm Certification Program (listed in https://charts.openshift.io/index.yaml).

File: test-network-function/identifiers/identifiers.go [hunk 5]
```diff
@@ -431,6 +436,15 @@ with no resourceNames under its rules.`),
...
+		Description: formDescription(TestHelmIsCertifiedIdentifier,
+			`tests whether CNF Helm listed in the cluster passed the Red Hat Helm Certification Program (OCP).`),
```
Comment (edcdavid) [inline test-network-function/identifiers/identifiers.go:444]: tests whether helm charts listed in the cluster passed the Red Hat Helm Certification Program.

File: internal/api/api.go [hunk 6]
```diff
@@ -148,6 +150,29 @@ type ContainerCatalogEntry struct {
...
 }
+type ChartStruct struct {
+	Entries map[string][]struct {
+		Name        string `json:"name"`
+		Version     string `json:"version"`
+		KubeVersion string `json:"kubeVersion"`
+	} `json:"entries"`
+}
+
+func convert(i interface{}) interface{} {
```
Comment (edcdavid) [inline internal/api/api.go:161]: rename convert -> convertYamlToJson

File: test-network-function/certification/suite.go [hunk 7]
```diff
@@ -68,9 +68,46 @@ var _ = ginkgo.Describe(common.AffiliatedCertTestKey, func() {
...
+		return certAPIClient.GetYamlFile()
+	}
+}*/
+func testHelmCertified(env *configpkg.TestEnvironment) {
+	testID := identifiers.XformToGinkgoItIdentifier(identifiers.TestHelmIsCertifiedIdentifier)
+	ginkgo.It(testID, ginkgo.Label(testID), func() {
+		certAPIClient = api.NewHTTPClient()
+		helmcharts := env.Helmcharts
+		if len(helmcharts) == 0 {
+			ginkgo.Skip("No helms to check  ")
```
Comment (edcdavid) [inline test-network-function/certification/suite.go:87]: "No helms to check"->"No helm charts to check"

File: test-network-function/certification/suite.go [hunk 8]
```diff
@@ -68,9 +68,46 @@ var _ = ginkgo.Describe(common.AffiliatedCertTestKey, func() {
...
+						log.Info(fmt.Sprintf("Helm %s with version %s is certified", helm.Chart, helm.Version))
+						break
+					}
+				}
+				if certified {
+					break
+				}
+			}
+			if !certified {
+				log.Info(fmt.Sprintf("Helm %s with version %s is not certified", helm.Chart, helm.Version))
```
Comment (edcdavid) [inline test-network-function/certification/suite.go:105]: Helm -> helm chart

File: test-network-function/identifiers/identifiers.go [hunk 9]
```diff
@@ -121,6 +121,11 @@ var (
 		Url:     formTestURL(common.AffiliatedCertTestKey, "operator-is-certified"),
 		Version: versionOne,
 	}
+	// TestHelmIsCertifiedIdentifier tests that helm chart has passed helm certification.
+	TestHelmIsCertifiedIdentifier = claim.Identifier{
+		Url:     formTestURL(common.AffiliatedCertTestKey, "helm-is-certified"),
```
Comment (edcdavid) [inline test-network-function/identifiers/identifiers.go:126]: helm-is-certified -> helmchart-is-certified

File: internal/api/api.go [hunk 10]
```diff
@@ -148,6 +150,29 @@ type ContainerCatalogEntry struct {
 		TopLayerID             string `json:"top_layer_id"`
 		UncompressedTopLayerID string `json:"uncompressed_top_layer_id"`*/
 }
+type ChartStruct struct {
+	Entries map[string][]struct {
+		Name        string `json:"name"`
+		Version     string `json:"version"`
+		KubeVersion string `json:"kubeVersion"`
```
Comment (edcdavid) [inline internal/api/api.go:157]: Why not use KubeVersion to check that the platform can run this helm chart :
[deliedit@fedora newrepo]$ oc version
Client Version: 4.8.13
Server Version: 4.8.3
Kubernetes Version: *v1.21.1*+051ac4f

File: pkg/config/autodiscover/helm_info.go [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+type HelmSetList struct {
+	Items []HelmList `json:""`
+}
+type HelmList struct {
+	Name      string `json:"name"`
+	Namespace string `json:"namespace"`
+	Chart     string `json:"chart"`
+}
+
+func Gethelm() *HelmSetList {
```
Comment (edcdavid) [inline pkg/config/autodiscover/helm_info.go:12]: Gethelm -> GetClusterHelmCharts

File: .github/workflows/pre-main.yaml [hunk 12]
```diff
@@ -155,7 +166,11 @@ jobs:
           working_directory: cnf-certification-test-partner
 
       # Perform smoke tests.
-
+      - name: wich helm
+        run: which helm
```
Comment (edcdavid) [inline .github/workflows/pre-main.yaml:170]: remove debug logs

File: .github/workflows/pre-main.yaml [hunk 13]
```diff
@@ -187,7 +202,10 @@ jobs:
           mkdir -p $TNF_CONFIG_DIR $TNF_OUTPUT_DIR
           cp test-network-function/*.yml $TNF_CONFIG_DIR
         shell: bash
-
+      - name: wich helm
+        run: which helm
+      - name: echo $PATH
```
Comment (edcdavid) [inline .github/workflows/pre-main.yaml:207]: same here

File: .github/workflows/pre-main.yaml [hunk 14]
```diff
@@ -133,8 +133,19 @@ jobs:
       - name: Install ginkgo
         run: go install github.com/onsi/ginkgo/ginkgo@v1.16.5
 
+      - name: install helm
+        run: git clone https://github.com/helm/helm.git
+
+      - name: make install helm
```
Comment (edcdavid) [inline .github/workflows/pre-main.yaml:139]: remove this section since helm is installed in the Dockerfile

**Comment by edcdavid** [comment]
Modify the title to remove CNFCERT and add that it contains also a bugfix for the container image generation

File: test-network-function/certification/suite.go [hunk 15]
```diff
@@ -240,3 +262,41 @@ func GetOcpVersion() string {
...
+		kubeVersion = strings.Split(kubeVersion, "+")[0]
+		kubeVersion = kubeVersion[1:]
+	} else {
+		kubeVersion = ""
+	}
+	return kubeVersion
+}
+func CompareNumber(num1, num2, kind string) bool {
+	newnum1 := strings.Split(num1, ".")
+	firstNum1, _ := strconv.ParseFloat(newnum1[0]+"."+newnum1[1], 64)
```
Comment (edcdavid) [inline test-network-function/certification/suite.go:278]: To compare version, maybe we can use a library, for instance:

There is a nice solution from Hashicorp - https://github.com/hashicorp/go-version

import github.com/hashicorp/go-version
v1, err := version.NewVersion("1.2")
v2, err := version.NewVersion("1.5+metadata")
// Comparison example. There is also GreaterThan, Equal, and just
// a simple Compare that returns an int allowing easy >=, <=, etc.
if v1.LessThan(v2) {
    fmt.Printf("%s is less than %s", v1, v2)
}

File: test-network-function/certification/suite.go [hunk 16]
```diff
@@ -44,9 +45,9 @@ const (
 )
 
 var (
-	ocpVersionCommand = "oc version -o json | jq '.openshiftVersion'"
-
-	execCommandOutput = func(command string) string {
+	ocpVersionCommand        = "oc version -o json | jq '.openshiftVersion'"
+	kubernatesVersionCommand = "oc version -o json | jq '.serverVersion.gitVersion'"
```
Comment (edcdavid) [inline test-network-function/certification/suite.go:49]: kubernatesVersionCommand -> kubernetesVersionCommand

File: Dockerfile [hunk 17]
```diff
@@ -97,6 +99,7 @@ RUN yum remove -y gcc git wget && \
 	rm -rf /usr/lib/golang/pkg && \
 	rm -rf /usr/lib/golang/src
 
+
```
Comment (hamadise) [inline Dockerfile:102]: non necessary line

File: Makefile [hunk 18]
```diff
@@ -129,6 +129,9 @@ install-tools:
 	go install github.com/onsi/ginkgo/v2/ginkgo@v2.1.2
 	go install github.com/onsi/gomega
 	go install github.com/golang/mock/mockgen@v1.6.0
+	git clone https://github.com/helm/helm.git && cd helm && make install
+
```
Comment (hamadise) [inline Makefile:133]: non necessary lines

File: internal/api/api.go [hunk 19]
```diff
@@ -227,12 +252,35 @@ func (api CertAPIClient) GetOperatorBundleIDByPackageName(org, name, vsersion st
...
 	responseData, err := api.getRequest(url)
 	if err == nil {
 		imageID, err = api.getIDFromResponse(responseData)
 	}
+
 	return imageID, err
 }
+func (api CertAPIClient) GetYamlFile() (ChartStruct, error) {
+	url := ("https://charts.openshift.io/index.yaml")
+	responseData, err := api.getRequest(url)
```
Comment (hamadise) [inline internal/api/api.go:265]: check here if err is not nil before proceeding

File: internal/api/api.go [hunk 20]
```diff
@@ -148,6 +150,29 @@ type ContainerCatalogEntry struct {
 		TopLayerID             string `json:"top_layer_id"`
 		UncompressedTopLayerID string `json:"uncompressed_top_layer_id"`*/
 }
+type ChartStruct struct {
+	Entries map[string][]struct {
+		Name        string `json:"name"`
+		Version     string `json:"version"`
+		KubeVersion string `json:"kubeVersion"`
+	} `json:"entries"`
+}
```
Comment (hamadise) [inline internal/api/api.go:159]: you can unmarshall yaml to this struct directly without passing by JSON transformation. 
``
yaml.Unmarshal()
``

File: internal/api/api.go [hunk 21]
```diff
@@ -227,12 +252,35 @@ func (api CertAPIClient) GetOperatorBundleIDByPackageName(org, name, vsersion st
...
+	var body interface{}
+	var charts ChartStruct
+	if errorr := yaml.Unmarshal(responseData, &body); errorr != nil && err != nil {
+		panic(errorr)
+	}
+
+	body = convertYamlToJSON(body)
+
+	if b, eror := json.Marshal(body); eror != nil {
+		panic(eror)
```
Comment (hamadise) [inline internal/api/api.go:275]: if you panic here the program will stop. 
you should log an error, mark the test as failed.

File: pkg/config/autodiscover/helm_info.go [hunk 22]
```diff
@@ -0,0 +1,22 @@
+package autodiscover
+
+type HelmSetList struct {
+	Items []HelmList `json:""`
+}
+type HelmList struct {
+	Name      string `json:"name"`
+	Namespace string `json:"namespace"`
+	Chart     string `json:"chart"`
```
Comment (hamadise) [inline pkg/config/autodiscover/helm_info.go:9]: I think ```HelmList ```  is not an appropriate name. This structure does not contain any list, it defined a single Helm object.

File: test-network-function/certification/suite.go [hunk 23]
```diff
@@ -68,9 +69,67 @@ var _ = ginkgo.Describe(common.AffiliatedCertTestKey, func() {
 
 		testContainerCertificationStatus()
 		testAllOperatorCertified(env)
+		testHelmCertified(env)
 	}
 })
 
+//
+/*func helmCertified() func() (interface{}, error) {
+	return func() (interface{}, error) {
```
Comment (hamadise) [inline test-network-function/certification/suite.go:78]: remove dead code.

File: test-network-function/certification/suite.go [hunk 24]
```diff
@@ -68,9 +69,67 @@ var _ = ginkgo.Describe(common.AffiliatedCertTestKey, func() {
...
+		certAPIClient = api.NewHTTPClient()
+		helmcharts := env.HelmchartsUnderTest
+		if len(helmcharts) == 0 {
+			ginkgo.Skip("No helm charts to check")
+		}
+		ourKubeVersion := GetKubeVersion()[1:]
+		out, _ := certAPIClient.GetYamlFile()
+		for _, helm := range helmcharts {
+			certified := false
+			for _, v := range out.Entries {
```
Comment (hamadise) [inline test-network-function/certification/suite.go:95]: if ```out``` never nil ?

File: test-network-function/certification/suite.go [hunk 25]
```diff
@@ -68,9 +69,71 @@ var _ = ginkgo.Describe(common.AffiliatedCertTestKey, func() {
...
 	}
 })
 
+//
+/*func helmCertified() func() (interface{}, error) {
+	return func() (interface{}, error) {
+		return certAPIClient.GetYamlFile()
+	}
+}*/
+func testHelmCertified(env *configpkg.TestEnvironment) {
```
Comment (hamadise) [inline test-network-function/certification/suite.go:82]: this test would never fail.
Also, I would say the logic is a  hard to follow. there's 7 levels of fors and ifs in it.
Can you add comments or make it a little bit simpler

File: Makefile (hunk 18 above)
Comment (aabughosh) [reply Makefile:133]: this is to install helm .. we need it for the test

File: test-network-function/certification/suite.go [hunk 26]
```diff
@@ -68,9 +69,56 @@ var _ = ginkgo.Describe(common.AffiliatedCertTestKey, func() {
...
+					log.Info(fmt.Sprintf("Helm %s with version %s is certified", helm.Name, helm.Version))
+					break
+				}
+			}
+			if !certified {
+				failedHelms = append(failedHelms, helm)
+			}
+		}
+		if len(failedHelms) > 0 {
+			log.Warnf("Helms that are not certified: %+v", failedHelms)
```
Comment (hamadise) [inline test-network-function/certification/suite.go:116]: this is not a warning, it's an error
``log.Error("Helms that are not certified: %+v", failedHelms)``
Also, please print it in the claim file as well
``tnf.ClaimFilePrint("Helms that are not certified: %+v", failedHelms)``

File: pkg/config/autodiscover/helm_info.go [hunk 27]
```diff
@@ -0,0 +1,22 @@
...
+	Items []HelmChart `json:""`
+}
+type HelmChart struct {
+	Name      string `json:"name"`
+	Namespace string `json:"namespace"`
+	Chart     string `json:"chart"`
+}
+
+func GetClusterHelmCharts() *HelmSetList {
+	out := execCommandOutput("helm list -A -o json")
```
Comment (edcdavid) [inline pkg/config/autodiscover/helm_info.go:13]: The test case should be restricted to the namespaces under test. We should also add a way to skip some helm charts, for instance adding a entry in the tnf_config.yml to skip some helm charts names in all namespace under test.

File: .github/workflows/pre-main.yaml [hunk 28]
```diff
@@ -155,7 +155,6 @@ jobs:
           working_directory: cnf-certification-test-partner
 
       # Perform smoke tests.
-
```
Comment (edcdavid) [inline .github/workflows/pre-main.yaml:158]: no changes in this file, revert ?

File: pkg/config/autodiscover/autodiscover_targets.go [hunk 29]
```diff
@@ -102,6 +102,54 @@ func FindTestTarget(labels []configsections.Label, target *configsections.TestTa
...
+				}
+				chartslist = append(chartslist, chart)
+			}
+		}
+	}
+	return chartslist
+}
+
+// func to check if the helm is exist on the no need to check list that are under the tnf_config.yml
+func checkifnoneedtocheck(helmName string, notcheckhelmlist []configsections.Notcheckhelmlist) bool {
```
Comment (edcdavid) [inline pkg/config/autodiscover/autodiscover_targets.go:125]: rename checkifnoneedtocheck -> ifSkipHelmChart
notcheckhelmlist -> skipHelmList

File: pkg/config/configsections/common.go [hunk 30]
```diff
@@ -72,6 +75,7 @@ type TestConfiguration struct {
 	CrdFilters []CrdFilter `yaml:"targetCrdFilters" json:"targetCrdFilters"`
 	// AcceptedKernelTaints
 	AcceptedKernelTaints []AcceptedKernelTaintsInfo `yaml:"acceptedKernelTaints,omitempty" json:"acceptedKernelTaints,omitempty"`
+	Notcheckhelmlist     []Notcheckhelmlist         `yaml:"notcheckhelmlist,omitempty" json:"notcheckhelmlist,omitempty"`
```
Comment (edcdavid) [inline pkg/config/configsections/common.go:78]: Notcheckhelmlist->SkipHelmChartList

File: pkg/config/autodiscover/autodiscover_targets.go [hunk 31]
```diff
@@ -102,6 +102,54 @@ func FindTestTarget(labels []configsections.Label, target *configsections.TestTa
...
+	for _, ch := range charts.Items {
+		if ns[ch.Namespace] {
+			if !checkifnoneedtocheck(ch.Name, notcheckhelmlist) {
+				name, version := getHelmNameVersion(ch.Chart)
+				chart := configsections.HelmChart{
+					Version: version,
+					Name:    name,
+				}
+				chartslist = append(chartslist, chart)
+			}
```
Comment (edcdavid) [inline pkg/config/autodiscover/autodiscover_targets.go:118]: add log to indicate that chart was skipped

File: test-network-function/tnf_config.yml [hunk 32]
```diff
@@ -19,3 +19,5 @@ certifiedoperatorinfo:
 acceptedKernelTaints:
   - module: vboxsf
   - module: vboxguest
+notcheckhelmlist:
```
Comment (edcdavid) [inline test-network-function/tnf_config.yml:22]: notcheckhelmlist->skipHelmChartList

File: .github/workflows/pre-main.yaml [hunk 33]
```diff
@@ -155,7 +155,6 @@ jobs:
           working_directory: cnf-certification-test-partner
```
Comment (hamadise) [inline .github/workflows/pre-main.yaml:155]: general comments here 

I think we should update README, and provide an example in tnf_config.yaml for a helm chart.
Also, It would be helpful to add unit test for the code that parses the helm chart configuration

**Review by hamadise — APPROVED** [review]

File: test-network-function/certification/suite.go [hunk 34]
```diff
@@ -68,9 +69,60 @@ var _ = ginkgo.Describe(common.AffiliatedCertTestKey, func() {
...
+				if certified {
+					log.Info(fmt.Sprintf("Helm %s with version %s is certified", helm.Name, helm.Version))
+					break
+				}
+			}
+			if !certified {
+				failedHelms = append(failedHelms, helm)
+			}
+		}
+		if len(failedHelms) > 0 {
```
Comment (edcdavid) [inline test-network-function/certification/suite.go:118]: typo

File: pkg/config/autodiscover/autodiscover_targets.go [hunk 35]
```diff
@@ -102,6 +102,55 @@ func FindTestTarget(labels []configsections.Label, target *configsections.TestTa
...
+		}
+	}
+	return chartslist
+}
+
+// func to check if the helm is exist on the no need to check list that are under the tnf_config.yml
+func checkifnoneedtocheck(helmName string, skipHelmChartList []configsections.SkipHelmChartList) bool {
```
Comment (edcdavid) [inline pkg/config/autodiscover/autodiscover_targets.go:125]: checkifnoneedtocheck -> isSkipHelmChart

File: pkg/config/autodiscover/autodiscover_targets.go [hunk 36]
```diff
@@ -102,6 +102,55 @@ func FindTestTarget(labels []configsections.Label, target *configsections.TestTa
...
+	if len(skipHelmChartList) == 0 {
+		return false
+	}
+	for _, helm := range skipHelmChartList {
+		if helmName == helm.Name {
+			log.Infof("Helm chart with name %s was skiped", helmName)
```
Comment (edcdavid) [inline pkg/config/autodiscover/autodiscover_targets.go:131]: skiped -> skipped

File: test-network-function/certification/suite.go (hunk 34 above)
Comment (aabughosh) [reply test-network-function/certification/suite.go:118]: failedHelms where is the typo?

Comment (edcdavid) [reply test-network-function/certification/suite.go:118]: failedHelms->failedHelmCharts

**Review by edcdavid — APPROVED** [review]
